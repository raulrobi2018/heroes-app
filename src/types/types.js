export const types = () => {
    // El texto "[auth]" es para indicar a qué pertenece esa acción
    login: "[auth] login",
    logout: "[auth] logout"
}