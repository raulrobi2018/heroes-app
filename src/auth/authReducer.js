import {types} from "../types/types";

export const authReducer = (state = {}, action) => {
    switch (action.type) {
        case types.login:
            return {
                // Retorna todo lo que venga en el payload
                ...action.payload,
                logged: true
            };
            break;

        case types.logout:
            return {
                logged: false
            };

        default:
            return state;
    }
};
