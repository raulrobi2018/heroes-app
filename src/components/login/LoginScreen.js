import React from "react";

// AquÃ­ extraemos la propiedad history que ya existe en el componente Route
// ya que es agregado por React
export const LoginScreen = ({history}) => {
    const handleLogin = () => {
        // history.push("/");
        history.replace("/");
    };

    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr />
            <button className="btn btn-primary" onClick={handleLogin}>
                Login
            </button>
        </div>
    );
};
