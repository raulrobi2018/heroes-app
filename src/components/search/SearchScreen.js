import React, {useMemo} from "react";
import queryString from "query-string";
import {useLocation} from "react-router-dom";
import {useForm} from "../../hooks/useForm";
import {HeroCard} from "../heroes/HeroCard";
import {getHeroesByName} from "../../selectors/getHeroesByName";

export const SearchScreen = ({history}) => {
    // Hook de react router
    const location = useLocation();
    console.log(queryString.parse(location.search));

    // De los parámetros de la url solo me interesa la query (q)
    //Se iguala q a vacío por si en la url no va y evitar error ya que devuelve undefined
    const {q = ""} = queryString.parse(location.search);

    const [formValues, {handleInputChange}] = useForm({
        searchText: q
    });

    //   Desestructuro searchText de formValues
    const {searchText} = formValues;

    // Utilizamos el useMemo ya que no queremos obtener los heroes cada vez que se renderiza
    // Solo se recargará si el id cambia, que se indica con el segundo parámetro del useMemo
    // que son las dependencias. En este caso heorId es el que depende del cambio
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (event) => {
        event.preventDefault();
        history.push(`?q=${searchText}`);
    };

    return (
        <div>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Find a hero..."
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <button
                            type="submit"
                            className="btn btn-outline-primary m-1"
                        >
                            Search
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {q === "" && (
                        <div className="alert alert-info">Search a hero</div>
                    )}

                    {q !== "" && heroesFiltered.length === 0 && (
                        <div className="alert alert-danger">
                            There isn't a hero with {q}
                        </div>
                    )}

                    {heroesFiltered.map((h) => (
                        <HeroCard key={h.id} {...h} />
                    ))}
                </div>
                <div></div>
            </div>
        </div>
    );
};
