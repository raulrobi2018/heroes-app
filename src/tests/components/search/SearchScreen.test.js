import {mount, configure} from "enzyme";

import {MemoryRouter, Route, Router} from "react-router-dom";
import {LoginScreen} from "../../../components/login/LoginScreen";

import "@testing-library/jest-dom";

//Esto es necesario para que no falle el mount en React 17
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import {AuthContext} from "../../../auth/AuthContext";
import {types} from "../../../types/types";
import {SearchScreen} from "../../../components/search/SearchScreen";
configure({adapter: new Adapter()});
// ------------------------------------------------------

describe("Testing SearchScreen component", () => {
    test("should display the component correctly with default values", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={SearchScreen}></Route>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");
    });

    test("should display Batman and the input with the queryString value", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={SearchScreen}></Route>
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
    });
});
