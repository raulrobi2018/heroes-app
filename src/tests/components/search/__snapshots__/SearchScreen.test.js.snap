// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchScreen component should display the component correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
    <div />
  </div>
</div>
`;
